### DISCUSSION TITLE
Vertical lines of cursors

### DISCUSSION THREAD

**Discussion by sambeau** [zed-industries/zed#6436] [Feature Requests]
I would like to propose a mode for creating vertical lines of cursors by clicking and dragging, similar to a feature in Sublime Text (with two improvements). It would be incredibly useful for formatting columns of data and repetitive code: CSV files, SVG files, HTML tags at the end of lines, large data structures in code. Note that it is subtly different to selecting a rectangle.

While in this mode:

1. Dragging a cursor vertically in a document cancels any cursors and creates a new line of cursors from the first click
2. Dragging beyond the end of any line creates a cursor at the end of that line
3. Any empty line, preferably also any line that is whitespace, is ignored and gets no cursor
4. Dragging horizontally while in this mode creates a text selection in any line that has a character in the starting row of the selection.
5. When selecting horizontally beyond the ends of lines, the selection becomes ragged, continuing to select characters in the longer lines.
6. All cursors are placed based on the column where the cursor is at the end of the drag
7. Text wrap is ignored

Note that (3) is different behaviour to Sublime Text. It makes the formatting of large files with empty lines possible.

Note that (7) is different behaviour to Sublime Text. It makes it possible to edit columns of text in lines wider than the screen. 

While I would like the option to be able to assign this behaviour as default, I'm sure that other people would prefer it banished to somewhere like alt+cmd+click+drag.

Notes on current behaviours:

Currently, option/alt+click places another cursor wherever you click. Each time you drag with the alt key pressed it will make a new selection from the start of the drag to the end of the drag. It places the cursor at the end of the drag. This is useful for careful editing, but very slow for data formatting, especially in large columns. You can get some of the behaviour proposed by carefully clicking each position you want a cursor. It also supports selecting beyond the end of a line. However, any existing cursors aren't canceled so you have to click your first cursor without the modifier key, then press it for all subsequent cursors.

shift+alt is very similar to the proposal, but subtly different. It allows you to create a rectangular selection from your current cursor position to the clicked position. However, to try to use it in the way suggested in this proposal, you have to place your cursor at the beginning of a selection and click where you would like your selection to end, and vertical line of cursors to appear. You can create a ragged right edge by starting to the left and pulling out to the right, you cannot start from the right unless your starting line is the longest in the selection. This mode ignores blank lines unless the cursor begins on column 0.

The proposed behaviour could be created from shift+alt by changing the following:
1. Cursors are cleared at the start of the drag
2. Drags can start or end at a column beyond the end of any line
3. Word wrap is ignored
4. Never put a cursor on an empty line, even if the drag starts on column 0

(Not that I am suggesting replacing shift+alt)

**Comment by sambeau** [comment]
Duplicate of issue #4352 (mostly)

**Comment by sambeau** [comment]
This was originally marked 'idea' but it seems clear now that it should have been an enhancement issue. Apologies for that, I've spent the last decade working on a system where we kept feature requests well away from bug reports.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
